### DISCUSSION TITLE
Iot Hub Connection Refused

### DISCUSSION THREAD

**Discussion by Shotariusii** [Azure-Samples/iot-middleware-freertos-samples#312] [Q&A]
Hello,
I  implement Azure Freertos Sdk into my device and supply the TLS and TCP/IP stacks. 
I connected to socket successfully and When Device first Tries to connect Azure IoT hub, MQTT returns 'Connection Refused'. 
as see Connack value got back from Server, is " Connection Refused : connection refused not authorised " . 
what can be a problem ?

**Comment by danewalton-msft** [comment]
Hi @Shotariusii 
It could be a few things, but with "not authorized", it's going to be related to the authentication of your device. 

This could be the device ID being wrong, or your SAS key or client certificate being wrong. Can you give more details on what kind of authentication you are using?

**Comment by Shotariusii** [comment]
Thanks for replay. 

I use Sim7600 for TLS TCP/IP stack.
I don't think Certificate problem because, i have connected to Server Socket sucssefully, (with certificate and also without certificate ( no_authentication) ). 
DEVICE_SYMMETRIC_KEY, hostname and Device id - are copied from Azure protal. 

So after Socket connection, when Sdk  tries first MQTT connection  i am getting what is wrote in my previous question.

**Reply by danewalton-msft** [reply]
Are you using the Connection String or Symmetric Key? You should be using the yellow highlighted section below and not the red.

![image](https://user-images.githubusercontent.com/36710865/211637060-1e2c0732-c199-4b1a-ae4e-4b2f37afcfd4.png)

**Reply by Shotariusii** [reply]
Exactly, the same Symmetric key is represented in connection string, but i just coppied Symetric Key.

**Reply by danewalton-msft** [reply]
And have you supplied new TLS/TCP code which isn't in our SDK?

If you'd like to, you can send the connection info to my email address in my profile to test (hub, device id, and sas key). Just know that you should roll the primary key over to a new value if you do that for security purposes.

**Comment by Shotariusii** [comment]
Okay i will send you a email

**Comment by Shotariusii** [comment]
Yes i supplied new TLS/TCP with Sim7600 module.

here is connection string :
HostName=rdsoftwareiothub.azure-devices.net;DeviceId=hardware-test-device-1;SharedAccessKey=PijADCJbscMCr0GTMJ1mH4dp/m9erzEyF2lDsYorUm8=
i extract all of that data from here.

**Reply by danewalton-msft** [reply]
I'm able to connect using those credentials. 

**NOTE you shared those publicly so you should roll the SharedAccessKey.**

Did you write your own HMAC function for the SAS key?

**Reply by ericwolz** [reply]
@Shotariusii  Your connection string about is still active. Please reset the device keys or delete this device.

**Comment by Shotariusii** [comment]
I use the library which was placed in sample code. (As i remember name
Mdbtls folder)
Actually there was just names for libraries, I could not find it and then i
download it

**Reply by danewalton-msft** [reply]
It could possibly be your time function which is not creating a valid Linux epoch time for the password. It's hard for me to tell since the credentials work for me on our sample.

**Comment by Shotariusii** [comment]
I have time function which returns time in milliseconds since MCU runs.

I know i have to use Unix epoch time in seconds, but i didnt think that
would be a problem.  Can be that a problem ?

**Reply by danewalton-msft** [reply]
As long as it's getting a valid epoch time, that part of the SAS token should be okay. Are you able to run the sample just fine using those credentials and running our Linux sample?

Also, you can reply right to this thread in Github discussions instead of creating a new thread each time.

**Comment by Shotariusii** [comment]
I am using freertos sdk, which runs on stm32 microcontroller.

Our software team member, who is working on web development side,  idk
exactly how, but he simulate device and successfuly connected to iot hub
with that credential

**Comment by Shotariusii** [comment]
Sorry i couldn't get it.

Can i use time, since MCU runs in milliseconds for Time function.

Or it has to be unix epoch time?

**Reply by danewalton-msft** [reply]
It must be epoch time

**Comment by Shotariusii** [comment]
Yes its works. Thanks for helping me.

One more question,  when i send cloud to device message, in twin JSON file
cloudtodevice variable increments, but i am getting on my device that
message, but in the twins JSON files the variable doesn't decrease and
stays same or just increments. And after some times it sets to 0.

This is bc i am hendling the message wrong way ?

**Reply by danewalton-msft** [reply]
Hmm I'm not sure. If the payload generation is adding the variable, then you should see the variable value in the JSON. If the payload is hardcoded, you won't see it changed.

**Comment by Shotariusii** [comment]
Hello Dane

As you remember  i am using Azure Freertos SDK  for communication to Iot
Hub.
I was wondering, Does this SDK, (mqtt or TLS/SSL layer)  guarantee that
data will be sent and received as its original form ?
or do i need to add CRC or similar method to guarantee data correctness?

**Reply by danewalton-msft** [reply]
![image](https://user-images.githubusercontent.com/36710865/215599312-640f2e3c-6a7f-4941-85b1-f3a96f8c154a.png)
